### PR TITLE
Fix flaky KafkaIndexTaskTest.

### PR DESCRIPTION
**Pull request by gianm**
The testRunTransactionModeRollback case had many race conditions. Most notably,
it would commit a transaction and then immediately check to see that the results
were *not* indexed. This is racey because it relied on the indexing thread being
slower than the test thread.

Now, the case waits for the transaction to be processed by the indexing thread
before checking the results.

### CODE REVIEW

**Review by paul-rogers — COMMENTED** [review]
The test improvements look good, thanks for fixing the test. Would be good to also get a review from someone familiar with the Kafka area under test.

File: extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTest.java [hunk 1]
```diff
@@ -3135,13 +3168,16 @@ public void testMultipleLinesJSONText() throws Exception
...
       String wellformed = toJsonString(true, "2049", "d2", "y", "10", "22.0", "2.0") +
-                     toJsonString(true, "2049", "d3", "y", "10", "23.0", "3.0");
+                          toJsonString(true, "2049", "d3", "y", "10", "23.0", "3.0");
 
       //multiple objects in one Kafka record but some objects are in ill-formed format
       //as a result, the whole ProducerRecord will be discarded
-      String illformed = "{\"timestamp\":2049, \"dim1\": \"d4\", \"dim2\":\"x\", \"dimLong\": 10, \"dimFloat\":\"24.0\", \"met1\":\"2.0\" }" +
-                     "{\"timestamp\":2049, \"dim1\": \"d5\", \"dim2\":\"y\", \"dimLong\": 10, \"dimFloat\":\"24.0\", \"met1\":invalidFormat }" +
-                     "{\"timestamp\":2049, \"dim1\": \"d6\", \"dim2\":\"z\", \"dimLong\": 10, \"dimFloat\":\"24.0\", \"met1\":\"3.0\" }";
+      String illformed =
```
Comment (kfaraz) [inline extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTest.java:3175]: Nit: malformed

**Review by kfaraz — APPROVED** [review]
LGTM. Minor nitpicks.

**Comment by abhishekagarwal87** [comment]
I had one comment. Please correct me if my understanding was wrong. there is an infinite loop that breaks only if the dev code works as expected. there should be some timeout in that loop. sometimes the test will run forever and travis will kill the whole job. That makes it harder to debug.

File: extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTest.java [hunk 2]
```diff
@@ -2572,7 +2566,9 @@ public void testRunTransactionModeRollback() throws Exception
             0,
             "sequence0",
             new SeekableStreamStartSequenceNumbers<>(topic, ImmutableMap.of(0, 0L), ImmutableSet.of()),
-            new SeekableStreamEndSequenceNumbers<>(topic, ImmutableMap.of(0, 13L)),
+
+            // End offset 13 is one after 12 real messages + 2 txn control messages (last seen message: offset 13).
+            new SeekableStreamEndSequenceNumbers<>(topic, ImmutableMap.of(0, 14L)),
```
[the remark this answers is not in the document]
Comment (gianm) [reply extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTest.java:2571]: The flow is slightly different: I added one additional transaction to the test. So there's one more control message.

**Comment by gianm** [comment]
> I had one comment. Please correct me if my understanding was wrong. there is an infinite loop that breaks only if the dev code works as expected. there should be some timeout in that loop. sometimes the test will run forever and travis will kill the whole job. That makes it harder to debug.

The tests are all annotated with `@Test(timeout = 60_000L)`, which I think will cause JUnit to interrupt the thread and stop the waiting after a minute. So, as far as I can tell, we won't have this running-forever issue in these particular test cases.

**Pull request merged**